### ISSUE TITLE
Potential loop in TimberCore - rethink purpose?

### ISSUE THREAD

**Issue by mgmartel** [timber/timber#478]
Hi @jarednova,

There seems to be a potential nasty loop in `TimberCore`: the `__get` method checks `$this->meta( $field )`. If the `meta` method does not exist, the magic method `__call` is called, which in turn calls `__get`. You see where this is headed?

Now, the `TimberCoreInterface` _does_ require a `meta` method, so all objects implementing `TimberCoreInterface` are good. The ones in trouble are the ones that 
- extend `TimberCore`
- don't implement `TimberCoreInterface`

That is:
- `TimberArchives`
- `TimberMenu`
- `TimberTheme`

`TimberSite` does implement `TimberCoreInterface`, but defines the `meta` method to simply call `__get` as well. So it's part of the list.

I think that the problem stems from the fact that TimberCore has 2 purposes:
1. Be a generic object wrapper for all-things-Timber, because it exposes handy `__get` and `__call` magic methods, as well as `import`, etc.
2. Be a wrapper for WordPress objects, where the `__get` and `__call` also interact with object meta.

My guess is that it's best to split the two purposes up into two objects. Something like:

```
-TimberCore (or TimberObject?)
--TimberWPObject (or something..)
```

Where the main addition of WP objects is that they all have object meta. Both classes should implement their own interface, so that objects that do not extend either of them can still 'plug in' my implementing the same interface.

What do you think?

**Comment by palmiak** [comment]
@mgmartel @connorjburton what is the status of this one? I see it has-patch, wasn't merged, but i has over 2 years.

**Comment by jarednova** [comment]
@palmiak I think this remains a live issue. The contradictions and confusion could be (relatively) easily sorted out by re-applying #486 to the current codebase. That said, I think we would want to target it at 2.x

**Comment by gchtr** [comment]
There’s not only a problem with non-existing methods, but with all inaccessible methods. In this part:

https://github.com/timber/timber/blob/3947bf995e30ff836e87354a6a0119bc47fd0afd/lib/Core.php#L38-L48

When `$field` is a the name of **a private method** of an object, then `if ( method_exists($this, $field) ) {` will create an infinite loop, because like mentioned, `__call()` will call `__get()` which will again call `__call()`.

And I would know an even easier method to solve this. In #1861, I added a `MetaInterface` interface that would remove any "knowledge" of meta stuff from `Core` and moves it to the classes that implement the `MetaInterface`. The only remaining thing that depends on anything «meta» in `Core` is now that call that creates an infinite loop.

**My proposal: Remove [dynamic properties](https://twig.symfony.com/doc/2.x/recipes.html#using-dynamic-object-properties)**. Don’t make it possible to access custom fields directly through `$post->custom_field` or `{{ post.custom_field }}`.

- It would also make it easier to work on #1823.
- The only benefit for dynamic properties to exist is because it’s a tiny bit easier to write than `$post->meta( 'custom_field' )` or `{{ post.meta('custom_field') }}`. But I think having `meta()` in your code makes it easier to understand where a value comes from. 
- I guess without dynamic properties we could even remove `__isset()` and `__get()`. And even the `CoreInterface` interface.
- It would prevent issues with code that probably shouldn’t be used but still works. Two examples:
	- `$image->src` instead of `$image->src()`: https://github.com/timber/timber/blob/2.x/tests/test-timber-image.php#L953
	- `$menu_item->name` instead of `$menu_item->name()`: https://github.com/timber/timber/blob/2.x/lib/MenuItem.php#L82

Less magic. Less workarounds.

Opinions? (Pinging @pascalknecht and @acobster …)

**Comment by acobster** [comment]
Less magic is almost always good IMO. I like the terseness of dynamic properties, but as with most magical code, the drawbacks outweigh the benefits here. I say remove 'em.

**Comment by jarednova** [comment]
@acobster @gchtr I've been wrestling with this one and here's where I think I am:

I've always loved the non-opinionated nature of being able to do `{{ post.my_field }}` just as the same as I can do `{{ post.title }}` which is why dynamic properties were in Timber from very early on. However, given the potential issues (infinite loops, ambiguity, etc.). I can see how this might no longer be the best approach.

I think it's important to note that `{{ post.my_field }}` in your Twig template is still the preferred approach, however this should be via a method created in a `Timber\Post` subclass. Thus, when I see...

```twig
<h3>{{ post.meta('my_field') }}</h3>
```
I know there's no special processing other than WP and plug-ins that have hooked into that meta field (ACF, etc.). However ...

```twig
<h3>{{ post.my_field }}</h3>
```
... tells me that the theme author HAS written some kind of special processing for this field and I should refer to the subclass for appropriate info.

Forgive what feels like an academic diversion, but this has been a part of the approach from day one so I want to square this change philosophically with what's come before. So in short...

- @gchtr do you have what you need to move forward with #1823 and #1861?
- @gchtr based on this conversation/decisions are we safe to close both #478 and #486 ?

**Comment by acobster** [comment]
@jarednova good point. Doesn't seem academic to me. :)

**What about this: remove the direct `__get()` call from `__call()`.** Something like this:

```php
	public function __call( $field, $args ) {
		if ( property_exists($this, $field) ) {
			return $this->field;
		}
		if ( method_exists($this, 'meta') && $meta_value = $this->meta($field) ) {
			return $this->$field = $meta_value;
		}
                return false;
	}
```

Notice that this is just part of what `__get()` does, but guarantees we don't create an infinite loop. (We could push this down into a helper method for reuse between `__call()` and `__get()` if we wanted to, but I think this is a perfectly acceptable amount of duplication.) The other thing that `__get()` does is check if `method_exists`...but by definition that doesn't make sense for `__call()` to do.

I think this is a lot simpler than removing a core piece of the syntax and all the documentation changes (and likely bug fixes) that would have to go along with that.

With regards to this PR, I'm not fundamentally opposed to making a clearer distinction between things classes that do or don't implement `meta()`, but this approach could be greatly simplified IMO. Happy to go into more detail if people feel like picking this PR back up.

**Comment by jarednova** [comment]
@acobster that seems like a less drastic solve. I hear "infinite loop" and think "why are we killing ourselves?" but if that will allow us to preserve lots of backwards compatibility and eliminate risks it would seem to cover all of the bases. Thus:

### `{{ post.meta('my_field`) }}`
This is the _preferred_ way to get custom data

### `{{ post.my_field }}`
This is the _preferred_ way to get custom data that is modified by a custom method. However, if that method doesn't exist, Timber will fallback to the equivalent `{{ post.meta('my_field'} }}` call.

It seems like you know the solution here better than anyone else. Are you good to implement this in a PR? I think between what you'd submit at #1861 (with `MetaInterface`) we could close-up this and #486

**Comment by acobster** [comment]
@jarednova yup, makes sense. I'll rework `__call()` in a separate PR.

**Comment by jarednova** [comment]
Hey @acobster just wanted to check in and see on your status here. Right now #1904 is pending based the final outcome. Have you had a chance to tear into this one?

**Comment by acobster** [comment]
Sorry @jarednova, not really. I've been spread pretty thin recently. I expect to have more time for this over the next couple Fridays, though. :)

**Comment by jarednova** [comment]
@acobster awesome, sounds good!

**Comment by acobster** [comment]
@jarednova @gchtr OK, I managed to make some headway today, but it looks like tests are failing in part because the fallback to magically returning dynamic properties like `name` are not set:

https://travis-ci.org/sitecrafting/timber/jobs/500625432

I'm inclined to define a `name()` method for `Timber\Term` (and maybe other `Core` subclasses). Or maybe just call the property instead of a magic method in those tests? Thoughts?

**Comment by gchtr** [comment]
@acobster I would change the tests to call the property instead of the magic method. I see that `title()` is the common method in Timber to get an object’s human-friendly name, so we could even use the `title()` method in the tests.

**Comment by jarednova** [comment]
@acobster agree with @gchtr. But it's actually a good reminder that, where appropriate, we should make those tests reflect the Twig syntax (since that's what someone should be interacting with 99% of the time). For example:

### Before
```php
function testGettingWithCategory() {
	$cat = $this->factory->term->create(array('name' => 'News', 'taxonomy' => 'category'));
	$pids = $this->factory->post->create_many(6);
	$cats = $this->factory->post->create_many(3, array('post_category' => array($cat)) );
	$cat_post = $this->factory->post->create(array('post_category' => array($cat)) );

	$cat_post = new TimberPost($cat_post);
	$this->assertEquals('News', $cat_post->category()->name());

	$posts = Timber::get_posts(array('category' => $cat));
	$this->assertEquals(4, count($posts));
}
```

### After
```php
function testGettingWithCategory() {
	$cat = $this->factory->term->create(array('name' => 'News', 'taxonomy' => 'category'));
	$pids = $this->factory->post->create_many(6);
	$cats = $this->factory->post->create_many(3, array('post_category' => array($cat)) );
	$cat_post = $this->factory->post->create(array('post_category' => array($cat)) );

	$cat_post = new TimberPost($cat_post);
	$str = Timber::compile_string('{{ post.category.name }}', array('post' => $cat_post));
	$this->assertEquals('News', $str);

	$posts = Timber::get_posts(array('category' => $cat));
	$this->assertEquals(4, count($posts));
}
```

**Comment by gchtr** [comment]
This was fixed in #2027/#2032.
